Frustum culling

lragnarsson opened the pull request:
Fairly naive implementation of frustum culling in world coordinates.
Created a Camera class.

lragnarsson committed:
Created a Camera class. Added functions to update and check against a culling frustum. Needs more testing.

lragnarsson committed:
Added bounding sphere generation after model loading. Frustum culling is now done in world coordinates and should function properly. Can probably be optimized quite a bit

lragnarsson committed:
fixup! Added bounding sphere generation after model loading. Frustum culling is now done in world coordinates and should function properly. Can probably be optimized quite a bit

lragnarsson committed:
Merge branch 'master' into frustum_culling

Conflicts:
	src/Model.cpp

lragnarsson commented:
To verify that it works you can add the following before return false; on line 41 in Camera.hpp:

std::cout << i << std::endl;

hansfilipelo commented on src/Input.cpp:
```diff
@@ -7,22 +7,22 @@ void init_input()
 }
 
-void handle_keyboard_input(Camera &camera, bool &loop)
+void handle_keyboard_input(Camera* camera, bool &loop)
```
Still not all on board on this "pointer is better" argument since one can pass a pointer wo changing anything - but I'm fine with the change

hansfilipelo commented on src/Model.cpp:
```diff
@@ -86,6 +86,19 @@ Model::Model(const std::string path, const GLuint shader_program,
+
+        this->rot_matrix = rot_matrix;
+        this->m2w_matrix = m2w_matrix;
+        this->bounding_sphere_radius = bounding_sphere_radius;
```
How do we derive this value (bounding sphere)?

hansfilipelo commented:
Needs to merge with master (since master has had updates since PR was created).  

Small change proposition, why is bounding_sphere an in-argument to constructor for model when there's a function which will be called to calculate the value? Shouldn't the code look something like: 

```
Model::Model(const std::string path, const GLuint shader_program,
  const glm::mat4 rot_matrix, const glm::mat4 m2w_matrix) {

    this->rot_matrix = rot_matrix;
    this->m2w_matrix = m2w_matrix;
    this->bounding_sphere_radius = this->generate_bounding_sphere();
```

Am I missing something?

lragnarsson committed:
Reverted back to using static allocation of Camera. Removed Model constructor with bounding radius. generate_bounding_sphere now called in constructor.

lragnarsson committed:
Merge branch 'master' into frustum_culling

Conflicts:
	src/main.cpp

lragnarsson commented:
Good catch on the constructor! Fixed that and reverted back to not using a pointer. Don't remember why I changed it. If we wanted the benefits we would just allocate it statically and pass the adress. But for now we use a reference.
